parse.py

```diff
@@ -0,0 +1,292 @@
+#-*- coding: utf-8 -*-
+
+import html5, string
+
+# Global variables
+_tags = None
+_domParser = None
+
+def _convertEncodedText(txt):
+	"""
+	Convert HTML-encoded text into decoded string.
+
+	The reason for this function is the handling of HTML entities, which is not
+	properly supported by native JavaScript.
+
+	We use the browser's DOM parser to to this, according to
+	https://stackoverflow.com/questions/3700326/decode-amp-back-to-in-javascript
+
+	:param txt: The encoded text.
+	:return: The decoded text.
+	"""
+	global _domParser
+
+	if _domParser is None:
+		_domParser = eval("new DOMParser")
+
+	dom = _domParser.parseFromString("<!doctype html><body>" + str(txt), "text/html")
+	return dom.body.textContent
+
+def _buildDescription():
+	"""
+	Generates a dictionary of all to the html5-library
+	known tags and their associated objects and attributes.
+	"""
+	tags = {}
+
+	for cname in dir(html5):
+		cl = getattr(html5, cname)
+		try:
+
+			inst = cl()
+			if cname != "Body" and isinstance(inst, html5.Widget):
+				attr = []
+
+				for fname in dir(inst):
+					if fname.startswith("_set"):
+						attr.append(fname[4:].lower())
+
+				tags[cname.lower()] = (cl, attr)
+
+		except:
+			pass
+
+	for tag in sorted(tags.keys()):
+		print("%s: %s" % (tag, ", ".join(sorted(tags[tag][1]))))
+
+	return tags
+
+def fromHTML(html, appendTo = None, bindTo = None):
+	"""
+	Parses the provided HTML code according to the objects defined in the html5-library.
+
+	Constructs all objects as DOM nodes. The first level is chained into root.
+	If no root is provided, root will be set to html5.Body().
+
+	The HTML elements are parsed for notations of kind [name]="ident", making
+	the corresponding instance available to the widget as widget.ident in the
+	Python code.
+
+	Example:
+
+	```python
+	import html5
+
+	div = html5.Div()
+	html5.parse.fromHTML('''
+		<div>Yeah!
+			<a href="hello world" [name]="myLink" class="trullman bernd" disabled>
+			hah ala malla" bababtschga"
+			<img src="/static/images/icon_home.svg" style="background-color: red;"/>st
+			<em>ah</em>ralla <i>malla tralla</i> da
+			</a>lala
+		</div>''', div)
+
+	div.myLink.appendChild("appended!")
+	```
+	"""
+
+	def scanWhite(l):
+		"""
+		Scan and return whitespace.
+		"""
+
+		ret = ""
+		while l and l[0] in string.whitespace:
+			ret += l.pop(0)
+
+		return ret
+
+	def scanWord(l):
+		"""
+		Scan and return a word.
+		"""
+
+		ret = ""
+		while l and l[0] not in string.whitespace + "<>=\"'":
+			ret += l.pop(0)
+
+		return ret
+
+	global _tags
+	stack = []
+
+	# Obtain tag descriptions
+	if _tags is None:
+		_tags = _buildDescription()
+
+	# Handle defaults
+	if appendTo is None:
+		appendTo = html5.Body()
+
+	if bindTo is None:
+		bindTo = appendTo
+
+	# Prepare stack and input
+	stack.append((appendTo, None))
+	html = [ch for ch in html]
```

Comment:
General question: what is faster in python? [y for y in x] or x[:] or list(x) to obtain a shallow copy?

Reply:
The way it works in Python  may differ here because we compile into JavaScript. In my opinion no requirement to be changed now.